libraries/standard/http/src/http_client.c

```diff
@@ -1,42 +1,311 @@
+                         size_t valueLen )
 {
-    return HTTP_NOT_SUPPORTED;
+    HTTPStatus_t status = HTTP_INTERNAL_ERROR;
+    uint8_t * pBufferCur = pRequestHeaders->pBuffer;
+
+    /* (a) Check if there is enough space in buffer for additional header.
+     *     The additional "\r\n" at the end is used for checking that there
+     *     is a enough space for the last line of an HTTP header.
+     *     This last line must be added separately after this method returns. */
+    size_t toAddLen = fieldLen + HTTP_HEADER_FIELD_SEPARATOR_LEN + \
+                      valueLen + HTTP_HEADER_LINE_SEPARATOR_LEN +  \
+                      HTTP_HEADER_LINE_SEPARATOR_LEN;
+
+    if( pRequestHeaders->headersLen
+        + toAddLen > pRequestHeaders->bufferLen )
+    {
+        /* TODO: Add log. */
+        return HTTP_INSUFFICIENT_MEMORY;
+    }
+
+    /* Write "Field: Value \r\n" to headers. */
+    memcpy( pBufferCur, pField, fieldLen );
+    pBufferCur += fieldLen;
+    memcpy( pBufferCur, HTTP_HEADER_FIELD_SEPARATOR,
+            HTTP_HEADER_FIELD_SEPARATOR_LEN );
+    pBufferCur += HTTP_HEADER_FIELD_SEPARATOR_LEN;
+    memcpy( pBufferCur, pValue, valueLen );
+    pBufferCur += valueLen;
+    memcpy( pBufferCur, HTTP_HEADER_LINE_SEPARATOR, HTTP_HEADER_LINE_SEPARATOR_LEN );
+
+    /* Subtract HTTP_HEADER_LINE_SEPARATOR_LEN because it is not actually written
+     * and only used for error checking as mentioned above in (a). */
+    pRequestHeaders->headersLen += toAddLen - HTTP_HEADER_LINE_SEPARATOR_LEN;
+
+    return status;
 }
 
-HTTPStatus_t HTTPClient_AddRangeHeader( HTTPRequestHeaders_t *pRequestHeaders,
-                                            int32_t rangeStart,
-                                            int32_t rangeEnd )
+
+HTTPStatus_t HTTPClient_InitializeRequestHeaders( HTTPRequestHeaders_t * pRequestHeaders,
+                                                  const HTTPRequestInfo_t * pRequestInfo )
 {
-    return HTTP_NOT_SUPPORTED;
+    HTTPStatus_t status = HTTP_INTERNAL_ERROR;
+    size_t toAddLen = 0;
+    uint8_t * pBufferCur = pRequestHeaders->pBuffer;
+    size_t httpsProtocolVersionLen = STRLEN_LITERAL( HTTP_PROTOCOL_VERSION );
+
+    pRequestHeaders->headersLen = 0;
+    /* Clear user-provided buffer. */
+    memset( pRequestHeaders->pBuffer, 0, pRequestHeaders->bufferLen );
+
+    /* Check for null parameters. */
+    if( _isNullParam( pRequestHeaders ) || _isNullParam( pRequestInfo ) ||
+        _isNullParam( pRequestHeaders->pBuffer ) ||
+        _isNullParam( pRequestInfo->method ) ||
+        _isNullParam( pRequestInfo->pHost ) ||
+        _isNullParam( pRequestInfo->pPath ) )
+    {
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    /* Check if buffer can fit "<METHOD> <PATH> HTTP/1.1\r\n". */
+    toAddLen = pRequestInfo->methodLen +                 \
+               SPACE_CHARACTER_LEN +                     \
+               pRequestInfo->pathLen +                   \
```

Comment:
If path was null and we added the "/" we will want to account for that? I think if unit tests are complete you will have caught this.